File: leetcode/1502. Can Make Arithmetic Progression From Sequence/main.ex
defmodule Solution do
  @moduledoc """
  Checks for an arithmetic progression by sorting all elements (a valid AP must be in sorted order), grabbing each pair of elements, mapping them to their difference, and ensuring all subtractions lead to the same difference by filtering the list to unique values.
  """

  @spec can_make_arithmetic_progression(arr :: [integer]) :: boolean
  def can_make_arithmetic_progression(arr) do
    Enum.sort(arr)
    |> Enum.chunk_every(2, 1, :discard)
    |> Enum.map(fn ([a, b]) -> b - a end)
    |> Enum.uniq()
    |> Enum.count()
    == 1
  end
end

File: leetcode/1502. Can Make Arithmetic Progression From Sequence/main.rs
impl Solution {
    pub fn can_make_arithmetic_progression(arr: Vec<i32>) -> bool {
        let mut arr = arr.to_vec();
        arr.sort();
        let mut diff = None;
        // simply iterate the sorted array and check if the difference between consecutive elements is the same
        for i in 0..(arr.len() - 1) {
            match diff {
                None => {
                    diff = Some(arr[i + 1] - arr[i]);
                }
                Some(d) => {
                    diff = Some(arr[i + 1] - arr[i]);
                    if diff.unwrap() != d {
                        return false;
                    }
                }
            }
        }
        true
    }
}